contracts/liquidity_hub/pool-network/terraswap_pair/src/helpers.rs

```diff
@@ -5,44 +5,345 @@ use cosmwasm_schema::cw_serde;
+            let return_amount: Uint256 =
+                return_amount - swap_fee_amount - protocol_fee_amount - burn_fee_amount;
 
-    Ok(SwapComputation {
-        return_amount: return_amount.try_into()?,
-        spread_amount: spread_amount.try_into()?,
-        swap_fee_amount: swap_fee_amount.try_into()?,
-        protocol_fee_amount: protocol_fee_amount.try_into()?,
-        burn_fee_amount: burn_fee_amount.try_into()?,
-    })
+            Ok(SwapComputation {
+                return_amount: return_amount
+                    .try_into()
+                    .map_err(|_| ContractError::SwapOverflowError {})?,
+                spread_amount: spread_amount
+                    .try_into()
+                    .map_err(|_| ContractError::SwapOverflowError {})?,
+                swap_fee_amount: swap_fee_amount
+                    .try_into()
+                    .map_err(|_| ContractError::SwapOverflowError {})?,
+                protocol_fee_amount: protocol_fee_amount
+                    .try_into()
+                    .map_err(|_| ContractError::SwapOverflowError {})?,
+                burn_fee_amount: burn_fee_amount
+                    .try_into()
+                    .map_err(|_| ContractError::SwapOverflowError {})?,
+            })
+        }
+        PairType::StableSwap { amp } => {
+            let offer_pool = Decimal256::decimal_with_precision(offer_pool, offer_precision)?;
+            let ask_pool = Decimal256::decimal_with_precision(ask_pool, ask_precision)?;
+            let offer_amount = Decimal256::decimal_with_precision(offer_amount, offer_precision)?;
+
+            let new_pool = calculate_stableswap_y(
+                offer_pool,
+                ask_pool,
+                offer_amount,
+                amp,
+                offer_precision,
+                ask_precision,
+                StableSwapDirection::Simulate,
+            )?;
+            let return_amount = ask_pool
+                .to_uint256_with_precision(u32::from(ask_precision))?
+                .checked_sub(Uint256::from_uint128(new_pool))?;
+
+            // the spread is the loss from 1:1 conversion
+            // thus is it the offer_amount - return_amount
+            let spread_amount = offer_amount
+                .to_uint256_with_precision(u32::from(ask_precision))?
+                .checked_sub(return_amount)?;
+
+            // subtract fees from return_amount
+            let swap_fee_amount: Uint256 = pool_fees.swap_fee.compute(return_amount);
+            let protocol_fee_amount: Uint256 = pool_fees.protocol_fee.compute(return_amount);
+            let burn_fee_amount = pool_fees.burn_fee.compute(return_amount);
+
+            let return_amount = return_amount
+                .checked_sub(swap_fee_amount)?
+                .checked_sub(protocol_fee_amount)?
+                .checked_sub(burn_fee_amount)?;
+
+            Ok(SwapComputation {
```

Comment:
This `Ok(SwapComputation)` is duplicated with by the `PairType::ConstantProduct` case. What if we return it only once outside of the match statement in L258? from the match we can either return a tuple with all the values we need to build the `SwapComputation` struct or  just define variables outside of the match and then assign the right values within the cases? What do you think about that?

Reply:
My main gripe with that is that we would probably depend on the order of the tuple. Can definitely move the `Ok` part outside the `match` statement, but by having each branch return a `SwapComputation` it would be very obvious if we assigned variables to the wrong component on the struct (e.g., imagine we create a new branch and accidentally swap the burn fee and protocol fee order)

Reply:
Yeah sounds reasonable, let's leave it as it is then.